mesonbuild/modules/qt5.py

```diff
@@ -87,15 +87,16 @@ def _detect_tools(self, env, method):
         self.tools_detected = True
 
     def parse_qrc(self, state, fname):
-        abspath = os.path.join(state.environment.source_dir, state.subdir, fname)
```

Comment:
Tests are failing because fname can be both File and str. You should have conditional code here based on whether it's str or File. Also, for File you should use File.absolute_path().

Reply:
On other option would be to implement __repr__ in File object and just do str(fnane). This would work for both str and File.

Reply:
The problem with that is that there is no 'canonical representation' of a file object. It is context-dependent whether you want a path relative to the build directory, relative to a particular subdirectory, or an absolute path. An absolute path should be used as a last resort.

Reply:
Thanks, will fix this evening.

Reply:
Actually, File already has __str__, which returns relative_name. That should work just fine. https://github.com/mesonbuild/meson/blob/7de61aac126e3ecc01cdb1dd1f8f07fbcd887edc/mesonbuild/mesonlib.py#L134

Reply:
Yes, I added `__str__` for File, but don't use that please. Use `relative_name()` explicitly.

Reply:
Sorry, haven't had time to get back to this yet. Will have time in the coming week.

@nirbheek Ok, no prob. Do you think `relative_name` or `absolute_path` (in which case I'll need to look to see where build path and source path are available) is better?